### PR TITLE
[FLINK-22767][coordination] Optimize the initialization of LocalInputPreferredSlotSharingStrategy

### PR DESCRIPTION
**Pull request by Thesharing**
## What is the purpose of the change

*This pull request aims to optimize the initialization of `LocalInputPreferredSlotSharingStrategy`. This will make its computation complexity decrease from O(N^2) to O(N). For more details please visit FLINK-22767.*

## Brief change log

  - *Optimize the initialization of LocalInputPreferredSlotSharingStrategy*

## Verifying this change

- *Two unit tests added, which covers two scenarios: 1. Two vertices connected with an ALL-TO-ALL edge; 2.Two vertices connected with two JobEdges.*
- *Manually verified the change by running an end-to-end test. The job has two JobVertices. One is source, while the other is sink. The parallelism is 8k. Both the all-to-all edge and pointwise edge are tested. Both streaming mode and batch mode are tested.*
- *Manually verified with TPCDS.*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (**yes** / no / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/LocalInputPreferredSlotSharingStrategy.java [hunk 1]
```diff
@@ -214,87 +245,87 @@ private ExecutionSlotSharingGroup tryFindAvailableProducerExecutionSlotSharingGr
...
+                    executionSlotSharingGroups.computeIfAbsent(
+                            slotSharingGroup.getSlotSharingGroupId(), k -> new ArrayList<>());
+
+            final ExecutionSlotSharingGroup newGroup = new ExecutionSlotSharingGroup();
+            newGroup.setResourceProfile(slotSharingGroup.getResourceProfile());
+
+            correspondingGroups.add(newGroup);
+
+            // Once a new ExecutionSlotSharingGroup is created, it's available for all JobVertices
+            // in this  SlotSharingGroup
```
Comment (Thesharing) [inline flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/LocalInputPreferredSlotSharingStrategy.java:321]: Remove one space here.

File: flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/LocalInputPreferredSlotSharingStrategy.java [hunk 2]
```diff
@@ -94,12 +97,33 @@ public LocalInputPreferredSlotSharingStrategy create(
...
-        private final Map<ExecutionSlotSharingGroup, Set<JobVertexID>> assignedJobVerticesForGroups;
+        /**
+         * A JobVertex only belongs to a single {@link SlotSharingGroup}. A SlotSharingGroup is
+         * corresponding to a set of {@link ExecutionSlotSharingGroup}s. Thus, we can maintain
+         * available ExecutionSlotSharingGroups for each JobVertex.
+         *
+         * <p>Once an ExecutionSlotSharingGroup is created, it becomes available for all JobVertices
+         * in the corresponding SlotSharingGroup in the beginning.
+         *
+         * <p>Once an SchedulingExecutionVertex is added to the ExecutionSlotSharingGroup, the group
```
Comment (Thesharing) [inline flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/LocalInputPreferredSlotSharingStrategy.java:114]: a

File: flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/LocalInputPreferredSlotSharingStrategy.java [hunk 3]
```diff
@@ -214,87 +245,87 @@ private ExecutionSlotSharingGroup tryFindAvailableProducerExecutionSlotSharingGr
...
         private boolean inSameLogicalSlotSharingGroup(
-                final ExecutionVertexID executionVertexId1,
-                final ExecutionVertexID executionVertexId2) {
+                final JobVertexID jobVertexId1, final JobVertexID jobVertexId2) {
 
             return Objects.equals(
-                    getSlotSharingGroup(executionVertexId1).getSlotSharingGroupId(),
-                    getSlotSharingGroup(executionVertexId2).getSlotSharingGroupId());
+                    checkNotNull(slotSharingGroupMap.get(jobVertexId1)).getSlotSharingGroupId(),
+                    checkNotNull(slotSharingGroupMap.get(jobVertexId2)).getSlotSharingGroupId());
```
[the remark this answers is not in the document]
Comment (Thesharing) [reply flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/LocalInputPreferredSlotSharingStrategy.java]: Thank you for pointing this out. Resolved.

File: flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/LocalInputPreferredSlotSharingStrategy.java [hunk 4]
```diff
@@ -214,87 +245,87 @@ private ExecutionSlotSharingGroup tryFindAvailableProducerExecutionSlotSharingGr
...
-                        executionSlotSharingGroupMap.get(producerVertexId);
-
-                checkState(producerGroup != null);
-                if (isGroupAvailableForVertex(producerGroup, executionVertexId)) {
-                    return producerGroup;
+            for (ConsumedPartitionGroup consumedPartitionGroup :
+                    executionVertex.getConsumedPartitionGroups()) {
+
+                Iterator<ExecutionSlotSharingGroup> availableGroupIterator =
+                        getAvailableGroupsForConsumedPartitionGroup(
```
[the remark this answers is not in the document]
Comment (Thesharing) [reply flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/LocalInputPreferredSlotSharingStrategy.java]: Yes. It seems confusing. I've tried to rewrite these usages.

File: flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/LocalInputPreferredSlotSharingStrategy.java [hunk 5]
```diff
@@ -159,6 +184,12 @@ private ExecutionSlotSharingGroupBuilder(
             return executionSlotSharingGroupMap;
         }
 
+        /**
+         * The vertices are ordered since {@link
+         * org.apache.flink.runtime.scheduler.adapter.DefaultExecutionTopology#getVertices} are
+         * ordered, i.e., {@link DefaultExecutionGraph#getAllExecutionVertices} are topologically
```
[the remark this answers is not in the document]
Comment (Thesharing) [reply flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/LocalInputPreferredSlotSharingStrategy.java]: Resolved.

File: flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/LocalInputPreferredSlotSharingStrategy.java [hunk 6]
```diff
@@ -94,12 +98,34 @@ public LocalInputPreferredSlotSharingStrategy create(
...
+        private final Map<JobVertexID, Set<ExecutionSlotSharingGroup>> availableGroupsForJobVertex;
+
+        /**
+         * Maintains the available {@link ExecutionSlotSharingGroup}s for every {@link
+         * ConsumedPartitionGroup}. The available groups are computed in {@link
+         * this#computeAllAvailableGroupsForConsumedPartitionGroup} when the ConsumedPartitionGroup
+         * is traversed for the first time.
+         */
+        private final Map<ConsumedPartitionGroup, Set<ExecutionSlotSharingGroup>>
+                availableProducerGroupsForConsumedPartitionGroup;
```
[the remark this answers is not in the document]
Comment (Thesharing) [reply flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/LocalInputPreferredSlotSharingStrategy.java]: I've tried to improve the doc.

File: flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/LocalInputPreferredSlotSharingStrategy.java [hunk 7]
```diff
@@ -214,87 +253,112 @@ private ExecutionSlotSharingGroup tryFindAvailableProducerExecutionSlotSharingGr
...
         private boolean inSameLogicalSlotSharingGroup(
-                final ExecutionVertexID executionVertexId1,
-                final ExecutionVertexID executionVertexId2) {
+                final JobVertexID jobVertexId1, final JobVertexID jobVertexId2) {
 
             return Objects.equals(
-                    getSlotSharingGroup(executionVertexId1).getSlotSharingGroupId(),
-                    getSlotSharingGroup(executionVertexId2).getSlotSharingGroupId());
+                    checkNotNull(getSlotSharingGroup(jobVertexId1)).getSlotSharingGroupId(),
+                    checkNotNull(getSlotSharingGroup(jobVertexId2)).getSlotSharingGroupId());
```
[the remark this answers is not in the document]
Comment (Thesharing) [reply flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/LocalInputPreferredSlotSharingStrategy.java]: Thank you for pointing this out. Resolved.

File: flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/LocalInputPreferredSlotSharingStrategyTest.java [hunk 8]
```diff
@@ -130,6 +142,43 @@ public void testInputLocalityIsRespected() {
...
     }
 
+    @Test
+    public void testInputLocalityIsRespectedWithAllToAllEdge() {
+        final TestingSchedulingTopology topology = new TestingSchedulingTopology();
+
+        final List<TestingSchedulingExecutionVertex> producer =
+                topology.addExecutionVertices()
+                        .withParallelism(2)
+                        .withJobVertexID(JOB_VERTEX_ID_1)
```
Comment (zhuzhurk) [inline flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/LocalInputPreferredSlotSharingStrategyTest.java:152]: Looks to me there is no need to explicitly set the `JOB_VERTEX_ID_1`.

**Review by zhuzhurk — COMMENTED** [review]
Thanks for addressing the comments! @Thesharing 
The change looks good to me except a few minor comments.
Please take a look.

**Review by zhuzhurk — APPROVED** [review]
LGTM.

**Comment by Thesharing** [comment]
Thank you for your review! It helped me a lot. The commits have been squashed. Waiting for the CI to give green.

**Pull request closed without merging**